src/OpenTelemetry/Logs/LogRecord.cs

```diff
@@ -88,18 +90,27 @@ public sealed class LogRecord
         /// langword="null"/> when <see
         /// cref="OpenTelemetryLoggerOptions.ParseStateValues"/> is enabled.
         /// </summary>
-        public object State { get; }
+        public object State { get; private set; }
 
         /// <summary>
         /// Gets the parsed state values attached to the log. Set when <see
         /// cref="OpenTelemetryLoggerOptions.ParseStateValues"/> is enabled
         /// otherwise <see langword="null"/>.
         /// </summary>
-        public IReadOnlyList<KeyValuePair<string, object>> StateValues { get; }
+        /// <remarks>
+        /// Note: StateValues are only available during the lifecycle of the log
+        /// message being written. If you need to capture state to be used later
+        /// (for example in batching scenarios), call <see cref="Buffer"/> to
```

Comment:
`Buffer` is an internal API and `StateValues` is public. Might lead to confusion being in the comment here. Should `Buffer` be a public API since maybe there are other use cases besides the batching processor where someone extending the SDK would want access to call `Buffer`?

Reply:
@alanwest I think it was public originally but then in code review on the initial PR we made it internal. Didn't catch the comments, sorry.

I took this in a different direction. Check out this commit: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2932/commits/c057ea0d963df313d2b1ead5bd0b0fc54938f619

The change yesterday to clear state after processing `LogRecord`s introduced/exposed an issue in our `InMemoryExporter`. Basically, it isn't very useful because the exported LogRecords don't have states or scopes available!

I added a setting on `SimpleLogRecordExportProcessor` which the `InMemoryExporter` log extension now uses to turn on buffering. It is currently public. If we leave it public, I think we can keep `LogRecord.Buffer` internal. Basically now if users need buffering they can either use `SimpleLogRecordExportProcessor` with the new property or `BatchLogRecordExportProcessor` which always buffers. I updated the comments referenced here to point to the public stuff.

/cc @cijothomas @reyang